src/main/java/org/gridsuite/sensitivityanalysis/server/service/nonevacuatedenergy/NonEvacuatedEnergyWorkerService.java

```diff
@@ -0,0 +1,936 @@
+                        GeneratorCapping capping = !monitoredBranchDetailResult.getGeneratorsCapping().containsKey(entry.getKey())
+                            ? entry.getValue() : monitoredBranchDetailResult.getGeneratorsCapping().get(entry.getKey());
+                        capping.setCumulatedCapping(capping.getCumulatedCapping() + entry.getValue().getCapping());
+                        monitoredBranchDetailResult.getGeneratorsCapping().put(entry.getKey(), capping);
+                        cappingsByEnergySource.putIfAbsent(generator.getEnergySource(), 0.);
+                        cappingsByEnergySource.put(generator.getEnergySource(), cappingsByEnergySource.get(generator.getEnergySource()) + capping.getCumulatedCapping());
+                    }
+                    monitoredBranchDetailResult.getCappingByEnergySource().putAll(cappingsByEnergySource);
+                    monitoredBranchDetailResult.setOverallCapping(cappingsByEnergySource.values().stream().mapToDouble(d -> d).sum());
+
+                    // keeping only the max computed variation
+                    if (!Double.isNaN(variation) && maxVariationForMonitoredBranch < variation) {
+                        maxVariationForMonitoredBranch = variation;
+                        maxGeneratorsCappings = new HashMap<>(generatorCappings.entrySet()
+                            .stream().collect(toMap(Map.Entry::getKey, e -> e.getValue().getCapping())));
+                    }
+
+                    mapEncounteredBranches.put(functionId, Boolean.TRUE);
+                }
+            }
+        }
+
+        if (maxVariationForMonitoredBranch < 0.3) {
+            // the max variation for all monitored branches is small :
+            // there is no limit violation detected and no further sensitivity analysis computation will be done for the current stage
+            noMoreLimitViolation.set(true);
+        } else {
+            // the max variation for all monitored branches is not small :
+            // there is a limit violation detected and a further sensitivity analysis computation will be done for the current stage, except if max iteration is reached
+            // we memorize the generators cappings needed to eliminate this limit violation
+            generatorsCappings.putAll(maxGeneratorsCappings);
+            noMoreLimitViolation.set(false);
+        }
+
+        return noMoreLimitViolation.get();
+    }
+
+    private void applyGeneratorsCappings(Network network, Map<String, Double> generatorsCappings) {
+        for (Map.Entry<String, Double> g : generatorsCappings.entrySet()) {
+            Generator generator = network.getGenerator(g.getKey());
+            if (generator != null) {
+                // we use the capping value to reduce the generator targetP
+                generator.setTargetP(generator.getTargetP() - g.getValue());
+            }
+        }
+    }
+
+    private void buildSummaryResults(NonEvacuatedEnergyResults nonEvacuatedEnergyResults) {
```

Comment:
each for loop could be extracted in separate method to make this method clearer?

Reply:
Refactored